### DISCUSSION TITLE
Can Testcontainers itself be used from inside a container?

### DISCUSSION THREAD

**Discussion by jernejg** [testcontainers/testcontainers-dotnet#423] [Q&A] [answered]
Hey, one question. I'm trying to run `dotnet test` from inside the container using:

`docker run -it -v /var/run/docker.sock:/var/run/docker.sock my:tag`

Found this suggestion here [Sibling docker containers](https://www.testcontainers.org/supported_docker_environment/continuous_integration/dind_patterns/)

But I get this exception
```
Error Message:
   DotNet.Testcontainers.Containers.ResourceReaperException : Initialization has been cancelled.
```

Is it even possible to run .net Testcontainers from inside docker?

**Comment by jernejg** [comment]
Hm looks like it works if I disable Ryuk
```
.WithCleanUp(false)
```

**Reply by HofmeisterAn** [reply]
It works with `WithCleanUp(true)` too. See my answer below. I do not recommend `WithCleanUp(false)` as a "fix".

**Comment by HofmeisterAn** [comment, marked as answer]
Yes, sibling containers are working. The `ResourceReaper` use it as well. Following example incl. running the default `ResourceReaper` in the background works fine:

```csharp
var testcontainers = new TestcontainersBuilder<TestcontainersContainer>()
  .WithImage("docker")
  .WithCommand(KeepTestcontainersUpAndRunning.Command)
  .WithBindMount("/var/run/docker.sock", "/var/run/docker.sock", AccessMode.ReadOnly)
  .Build();

await testcontainers.StartAsync()
  .ConfigureAwait(false);

var execResult = await testcontainers.ExecAsync(new[] { "docker", "ps" })
  .ConfigureAwait(false);
```

`execResult` contains the sibling containers:

```
CONTAINER ID   IMAGE                              COMMAND                  CREATED          STATUS                  PORTS                     NAMES
0d63ae7b2bb4   docker:latest                      "docker-entrypoint.s…"   1 second ago     Up Less than a second                             priceless_wu
e900901ee327   ghcr.io/psanetra/ryuk:2021.12.20   "/app"                   13 seconds ago   Up 12 seconds           0.0.0.0:59230->8080/tcp   testcontainers-ryuk-a815fb8c-f78f-4da9-a670-9ec706d96c07
```

----

```
Error Message:
   DotNet.Testcontainers.Containers.ResourceReaperException : Initialization has been cancelled.
```

This might happen if the default `ResourceReaper` does not start within 10 seconds. It's unusual, but maybe 10 seconds are not enough for your machine. Can you try to increase the default `initTimeout`?

https://github.com/HofmeisterAn/dotnet-testcontainers/blob/ffa5a812bde1b8e8de40e1849fe180b7e9b148fd/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L141

Unfortunately, you need to check out the repository. There is no option yet to change `initTimeout` for the default `ResourceReaper`.

**Reply by jernejg** [reply]
Thanks @HofmeisterAn I will look into it.

> This might happen if the default ResourceReaper does not start within 10 seconds. It's unusual, but maybe 10 seconds are not enough for your machine.

I have a brand new MacBook Pro with M1 laptop. I will try your suggestion, but I think it took less than 10 seconds after starting for this error to appear.

**Reply by HofmeisterAn** [reply]
> I have a brand new MacBook Pro with M1 laptop

That one should be fast enough 😂

> but I think it took less than 10 seconds after starting for this error to appear.

Sounds like something different. Can you try to pull and run [ryuk](https://hub.docker.com/r/testcontainers/ryuk) on your M1?

```
docker pull testcontainers/ryuk
docker run -v /var/run/docker.sock:/var/run/docker.sock -p 8080:8080 testcontainers/ryuk:latest
```

The above example runs fine at an intel Mac.

**Reply by HofmeisterAn** [reply]
@jernejg Does the error always occur?

**Reply by jernejg** [reply]
Hey @HofmeisterAn I haven't had the time to play around. Will do it this weekend and let you know.

**Reply by jernejg** [reply]
I noticed this:
Running `docker run -v /var/run/docker.sock:/var/run/docker.sock -p 8080:8080 testcontainers/ryuk:latest`
results in
![image](https://user-images.githubusercontent.com/5320517/154753182-f63751d5-fcea-419b-94ff-22249fbf0f0a.png)

But running `docker run -v /var/run/docker.sock:/var/run/docker.sock -p 8080:8080 ghcr.io/psanetra/ryuk:2021.12.20`
Doesn't produce the amd64 warning
![image](https://user-images.githubusercontent.com/5320517/154759075-3880e743-c1a3-4d0e-a929-8cae59368ba9.png)

**Reply by jernejg** [reply]
It does start though and looks like working ok, despite the warning:
```
docker run -v /var/run/docker.sock:/var/run/docker.sock -p 8080:8080 testcontainers/ryuk:latest
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
2022/02/18 21:11:43 Pinging Docker...
2022/02/18 21:11:43 Docker daemon is available!
2022/02/18 21:11:43 Starting on port 8080...
2022/02/18 21:11:43 Started!
```

**Reply by jernejg** [reply]
Nah it's not M1 specific. I managed to reproduce it on the build server. I'm doing something wrong, just don't know yet what :D

**Reply by HofmeisterAn** [reply]
> Nah it's not M1 specific

Ok, can you try the latest snapshot release `1.6.0-beta.2118`?

**Reply by jernejg** [reply]
I created minimum amount of code to replicate the issue https://github.com/jernejg/tc-demo

It is using the official Postgres test from the `testcontainers` readme file.

Running the test in a normal fashion using `dotnet test` works.

Running `run.sh` that builds an image and runs the tests inside:
```
docker build -t demo-test .
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock demo-test 
```

Using a Docker file:
```
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS builder
WORKDIR /src
COPY . .
CMD dotnet test "TestProject1.csproj"
```

Produces this error:
```
[xUnit.net 00:00:11.05]     TestProject1.PostgreSqlTest.ExecuteCommand [FAIL]
  Failed TestProject1.PostgreSqlTest.ExecuteCommand [1 ms]
  Error Message:
   DotNet.Testcontainers.Containers.ResourceReaperException : Initialization has been cancelled.
  Stack Trace:
     at DotNet.Testcontainers.Containers.ResourceReaper.GetAndStartNewAsync(Guid sessionId, String ryukImage, TimeSpan initTimeout, CancellationToken ct)
   at DotNet.Testcontainers.Containers.ResourceReaper.GetAndStartNewAsync(Guid sessionId, String ryukImage, TimeSpan initTimeout, CancellationToken ct)
   at DotNet.Testcontainers.Containers.ResourceReaper.GetAndStartDefaultAsync(CancellationToken ct)
   at DotNet.Testcontainers.Clients.TestcontainersClient.RunAsync(ITestcontainersConfiguration configuration, CancellationToken ct)
   at DotNet.Testcontainers.Containers.TestcontainersContainer.Create(CancellationToken ct)
   at DotNet.Testcontainers.Containers.TestcontainersContainer.StartAsync(CancellationToken ct)

```

There is still the possibility I'm doing sth wrong in the Dockerfile :worried:

**Reply by jernejg** [reply]
Reproduced on an Intel Windows machine running Ubuntu 18.04 via WSL

**Reply by HofmeisterAn** [reply]
Ok, your test is a little different from what I expected. In your case, .NET Testcontainer really fails to connect to Ryuk:

https://github.com/HofmeisterAn/dotnet-testcontainers/blob/0cf385a6d037504b2262332e58d4e32232b074d9/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L234

It's a bit tricky 😄. I just had a quick look, but in your case:

- Ryuk is created as sibling from your demo container
- The test cannot connect to `localhost:<mapped_port>` or to `<gateway_ip>:<mapped_port>`
- But it can connect to `<ryuk_container_ip>:8080` and to `host.docker.internal:<mapped_port>`

You will face the same problem when accessing the database later.
I created a branch `discussions/423` to simplify testing.

@PSanetra Do you have an idea how to improve it?

**Reply by PSanetra** [reply]
@HofmeisterAn I have created a PR which should probably fix this issue, although I could not reproduce the issue on my Manjaro Linux machine as the Ryuk ports are bound to localhost here: https://github.com/HofmeisterAn/dotnet-testcontainers/pull/426

**Reply by jernejg** [reply]
ah ok I think I get it. When it's not running in Docker containers communicate through the mapped port on the host.

But when you run in Docker, you can't use `localhost` you need to reference the container by name + they need to be on the same network for this right?

**Reply by HofmeisterAn** [reply]
> But when you run in Docker, you can't use `localhost` you need to reference the container by name + they need to be on the same network for this right?

Yes. But now, we just use the host (that maps the port) instead of the const `localhost`.

Can you try it again with `1.6.0-beta.2121`. My test runs fine.

**Reply by jernejg** [reply]
Looks like `1.6.0-beta.2121` fix the issues but broke something that was working before.

`.WithCleanUp(true)` now works on my MBP M1 with the sibling container strategy, however....

My CI pipeline started failing with a new error. This happens even if I use `.WithCleanUp(false)`.

If I downgrade to `1.6.0-beta.2118`, the issue is gone.

My CI pipeline runs on a Ubuntu VM `ubuntu-2004:202111-02`. I think `1.6.0-beta.2121` might have broken sth regarding IP's on Linux boxes.

The error (Postgres test from Readme.md):
```
[xUnit.net 00:00:05.33]     Adapter.EntityFramework.IntegrationTests.Tests.xxx [FAIL]
  Failed Adapter.EntityFramework.IntegrationTests.Tests.xxx [30 ms]
  Error Message:
   System.Net.NetworkInformation.PingException : An exception occurred during a Ping request.
---- System.Net.Internals.SocketExceptionFactory+ExtendedSocketException : Name or service not known
  Stack Trace:
     at System.Net.NetworkInformation.Ping.GetAddressAndSend(String hostNameOrAddress, Int32 timeout, Byte[] buffer, PingOptions options)
   at System.Net.NetworkInformation.Ping.Send(String hostNameOrAddress, Int32 timeout, Byte[] buffer, PingOptions options)
   at System.Net.NetworkInformation.Ping.Send(String hostNameOrAddress)
   at DotNet.Testcontainers.Containers.TestcontainersContainer.<>c.<GetContainerGateway>b__35_0(String hostName)
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at DotNet.Testcontainers.Containers.TestcontainersContainer.GetContainerGateway()
   at DotNet.Testcontainers.Containers.TestcontainersContainer.get_Hostname()
   at DotNet.Testcontainers.Containers.PostgreSqlTestcontainer.get_ConnectionString()
   at Adapter.EntityFramework.IntegrationTests.Tests.xxx() in /sln/test/integration/Adapter.EntityFramework.IntegrationTests/Tests/xxx:line 36
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
   at System.Net.Dns.GetHostEntryOrAddressesCore(String hostName, Boolean justAddresses)
   at System.Net.Dns.GetHostAddresses(String hostNameOrAddress)
   at System.Net.NetworkInformation.Ping.GetAddressAndSend(String hostNameOrAddress, Int32 timeout, Byte[] buffer, PingOptions options)
```

**Reply by HofmeisterAn** [reply]
We just need to catch the exception: https://github.com/HofmeisterAn/dotnet-testcontainers/blob/d095f96f325d90db398779fd93d1ee2e9d3a283c/src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs#L344

Sorry, for the back and forth. I thought that `Send` returns, always a status. I can do that later that day.

**Reply by HofmeisterAn** [reply]
`1.6.0-beta.2123` 😅?

**Reply by jernejg** [reply]
Hehe thanks!

So the results, if I run locally on my MBP M1 (normal + sibling) works fine.

In my CI (I always run in siblings mode there), the results are flaky. Sometimes it works and sometimes I get the `Initialization has been cancelled` error.  Timeout maybe?

I also noticed that when It's running in the CI pipeline (Linux), I can't use the `host.docker.internal` label to reference the host, when I replace it with `172.17.0.1` it works. But I think that might be a Docker+OS specific kind of a thing https://stackoverflow.com/questions/22944631/how-to-get-the-ip-address-of-the-docker-host-from-inside-a-docker-container

Thanks again for all the help.

**Reply by HofmeisterAn** [reply]
> Timeout maybe?

Yes, could be. The default [timeout](https://source.dot.net/#System.Net.Ping/System/Net/NetworkInformation/Ping.cs,15) seems to be 5 seconds for `Ping.Send`. On my tests, `Send` immediately fails, but maybe it's different on Linux. I can pass a timeout less than 5 seconds to `Send` in the future. Then you should be ok.

> when I replace it with `172.17.0.1` it works

Can you add `172.17.0.1  host.docker.internal` to `/etc/hosts`? Or set the timeout to a smaller value e.g. `ping.Send(hostName, 500)`:

https://github.com/HofmeisterAn/dotnet-testcontainers/blob/b99957969f1bf8d1615196f1a77899cb5d51c76e/src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs#L346

I'd like to make sure it's working before publishing another version.

**Reply by PSanetra** [reply]
I don't quite understand why the docker host ip is `172.17.0.1` and not `127.0.0.1`. Is the docker host a remote host?

**Reply by jernejg** [reply]
On Docker for Linux, the IP address of the gateway between the Docker host and the bridge network is 172.17. 0.1 if you are using default networking.

**Reply by PSanetra** [reply]
ok, in that case I am still not sure why it is not already working. I think the gateway ip should already be used as a fallback when `host.docker.internal` and `gateway.docker.internal` are not reachable via ping:
https://github.com/HofmeisterAn/dotnet-testcontainers/blob/b99957969f1bf8d1615196f1a77899cb5d51c76e/src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs#L355

**Reply by HofmeisterAn** [reply]
> I don't quite understand why the docker host ip is `172.17.0.1` and not `127.0.0.1`. Is the docker host a remote host?

That's because it's running DinD, if I'm not wrong.

```mermaid
graph LR
    A[Ubuntu VM<br/>127.0.0.1<br/>172.17.0.1]
    B[Demo<br/>172.17.0.2<br/>Executes tests];
    C[Ryuk<br/>172.17.0.3];
    D[PostgreSql<br/>172.17.0.4];
    A --> |Starts container| B;
    B --> |Starts container| C;
    B --> |Starts container| D;
```

> ok, in that case I am still not sure why it is not already working. I think the gateway ip should already be used as a fallback when `host.docker.internal` and `gateway.docker.internal` are not reachable via ping:

Those records do only exist in environments with Docker Desktop. In his example, both host names won't respond, it will fall back to: `this.container.NetworkSettings.Networks.First().Value.Gateway; `

**Reply by PSanetra** [reply]
@HofmeisterAn yes right, but should `this.container.NetworkSettings.Networks.First().Value.Gateway;` not already return the Gateway ip in the case that @jernejg mentioned if those records do not exist?

I am referring to this quote:

> I also noticed that when It's running in the CI pipeline (Linux), I can't use the host.docker.internal label to reference the host, when I replace it with 172.17.0.1 it works.

I think replacing `host.docker.internal` with `172.17.0.1` should not be necessary as `this.container.NetworkSettings.Networks.First().Value.Gateway;` should already return `172.17.0.1` if `host.docker.internal` does not exist. So I don't quite understand why the current code does not already work.

**Reply by HofmeisterAn** [reply]
Yes, you're right. It's not necessary, I think the current code works. My suggestion will just prevent the `Initialization has been cancelled.` (flaky test) exception if each ping has a timeout of 5 seconds (10 seconds in total).

**Reply by PSanetra** [reply]
Ah yes, a shorter timeout probably makes sense in that case. Maybe a dns lookup via [`Dns.GetHostEntry`](https://docs.microsoft.com/en-us/dotnet/api/system.net.dns.gethostentry) would also be faster. I am not sure if there are any tradeoffs using that api.

**Reply by HofmeisterAn** [reply]
I did some tests yesterday. It looks like both methods (`Dns.GetHostEntry`, `Ping.Send`) do not fail immediately or after the specific timeout. Both methods fail after approximately 2 seconds on my system if the host name does not exist. This may be different on other operating systems. It looks like the `Ping.Send` timeout refers to the ping response, not the command itself. I'll try different implementations in the next days.

**Reply by PSanetra** [reply]
Another idea might be to get the default gateway address in a similar way like in [testcontainers-java](https://github.com/testcontainers/testcontainers-java/blob/14eff07a7cae8d48b07b80e309bba601f164ae66/core/src/main/java/org/testcontainers/dockerclient/DockerClientConfigUtils.java#L23). ~Although I guess it should be possible to just use the dotnet api to get the default network interface and gateway ip as the tests are running anyway in a container, right? Maybe via [`NetworkInterface.GetAllNetworkInterfaces`](https://docs.microsoft.com/en-us/dotnet/api/system.net.networkinformation.networkinterface.getallnetworkinterfaces?view=net-6.0).~ (Edit: I forgot that the target container might run in a different network.)

**Reply by HofmeisterAn** [reply]
This should work:
https://github.com/HofmeisterAn/dotnet-testcontainers/blob/f25ed8705fa72c9d03eb185c41bd783db0fbb908/src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs#L355
